### PR TITLE
Fix POS checkout when products share the same title

### LINKED ISSUE #7527: [Bug]: POS checkout fails when products share the same display title
### What is your BTCPay version?

BTCPay Server v2.4.3+3f770d0ae45ffdcb609713f42d0b4412a4711c6e - Development (Debug)

### How did you deploy BTCPay Server?

Local development environment using the BTCPay Server repository, .NET, and the project's Docker Compose test infrastructure (Bitcoin Core regtest, NBXplorer, PostgreSQL, and supporting services).

### What happened?

POS checkout fails when two valid Point of Sale products have different product IDs but the same display title.

In my test, I created two fixed-price products with unique IDs (`variant-a` and `variant-b`) but gave both the display title `Variant`. Their prices were USD 1 and USD 2.

Both products could be added to the cart normally and the cart correctly showed a USD 3 total. However, clicking **Pay** caused checkout to fail with an unhandled exception instead of creating the invoice.

The exception is:

System.ArgumentException: An item with the same key has already been added. Key: Variant

The failure occurs in `BTCPayServer.Services.Invoices.PosReceiptData.UpdateFromCart(...)`, where the receipt data appears to use the product title as a dictionary key.

I reproduced the issue manually through the customer-facing POS UI and also with a focused integration regression test.

### How did you encounter this bug?

1. Created a store and connected the required wallets in a local BTCPay Server development environment.

2. Created a Point of Sale app using the **Product list with cart** style and USD as the currency.

3. Added the first fixed-price product:
   - Title: `Variant`
   - ID: `variant-a`
   - Price: USD 1

4. Added the second fixed-price product:
   - Title: `Variant`
   - ID: `variant-b`
   - Price: USD 2

5. Saved the POS app and opened the customer-facing POS using **View**.

6. Added one USD 1 `Variant` and one USD 2 `Variant` to the cart.

7. Confirmed that both items appeared separately and the cart total was correctly displayed as USD 3.

8. Clicked **Pay**.

9. Instead of creating a USD 3 invoice, BTCPay Server displayed an unhandled exception:

   `ArgumentException: An item with the same key has already been added. Key: Variant`

10. The stack trace points to:

   `BTCPayServer.Services.Invoices.PosReceiptData.UpdateFromCart(...)`

   in `BTCPayServer/Services/Invoices/PosReceiptData.cs`.

I repeated the scenario and received the same failure.

I also prepared a focused integration test using two products with distinct IDs and the same title. The test expects HTTP 200 and a USD 3 invoice but currently receives HTTP 500.

Image
Image

### Relevant log output

```shell
System.ArgumentException: An item with the same key has already been added. Key: Variant
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at BTCPayServer.Services.Invoices.PosReceiptData.UpdateFromCart(IEnumerable<AppItem> appItems, PosAppData jposData, String currency, DisplayFormatter displayFormatter)
```

### What browser do you use?

 Google Chrome (Chromium), Windows 10

### Additional information

The issue was reproduced both manually through the POS UI and through a focused integration test.

### Manual reproduction

The cart correctly contained:

- `Variant` — USD 1
- `Variant` — USD 2
- Total — USD 3

Clicking **Pay** then produced the duplicate-key exception instead of creating the invoice.

### Automated reproduction

A minimal local integration regression test was added temporarily for verification. It:

- creates two fixed-price products with distinct IDs;
- gives both products the title `Variant`;
- prices them at USD 1 and USD 2;
- adds one of each product to the cart;
- expects HTTP 200 and a USD 3 invoice.

The test failed identically twice because checkout returned HTTP 500.

Nearby normal POS integration tests continued to pass.

### Technical observation

`PosReceiptData.UpdateFromCart` appears to use the display title as the dictionary key for fixed-price receipt entries. Since both products have the title `Variant`, the second insertion results in a duplicate-key exception even though the underlying product IDs are different.

Current tested commit:

`3f770d0ae45ffdcb609713f42d0b4412a4711c6e`

### Are you sure this is a bug report?

- [x] I confirm this is a bug report

### PR DESCRIPTION
**Pull request by comwanga**
Fixes #7527.

Point of Sale checkout now supports purchasing distinct products that share the same display title.

Previously, adding both products to the cart worked, but clicking **Pay** failed during receipt construction because both products attempted to use the same receipt label.

Receipt labels are now disambiguated only when a collision occurs, first by unit price and then, if still necessary, by product ID. Products with unique titles keep their existing receipt labels unchanged.

For example, two products named `Variant`, priced at USD 1 and USD 2, now remain separate receipt entries and produce the correct USD 3 invoice.

A regression test covers:

- successful invoice creation;
- the USD 3 total;
- both receipt entries;
- the ordered receipt labels:
  - `Variant`
  - `Variant ($2.00)`

Manual end-to-end verification was also completed using the local regtest POS fixture. The USD 3 checkout succeeded, a Bitcoin invoice was created, and payment was received successfully with no duplicate-key exception.

#### Cart containing both same-title products

pos1

#### Successful payment

image

#### Receipt
image

### CODE REVIEW

**Comment by NicolasDorier** [comment]
Thanks. It indeed shouldn't crash, but why do you need variants? Is it just a feature you really need, or is it just a bug you encountered and you don';t want the crash?

**Comment by comwanga** [comment]
I don’t specifically need variants as a new feature. I encountered the HTTP 500 while testing a valid POS configuration where multiple products had the same display title.

The variant cases were added only as regression coverage because the existing receipt-key logic uses the price and item ID to disambiguate duplicate titles. I wanted to ensure that this logic cannot produce another duplicate dictionary key through a collision chain.

So the intended scope of this PR is simply: valid carts should not crash when receipt labels collide, while preserving the existing receipt-key format where possible.

**Comment by NicolasDorier** [comment]
@comwanga if that's the case, just replace the `Add` by `TryAdd`. Your code is difficult to review, and nobody really care about showing several lines with the same title. Just changing Add by TryAdd will ignore items with the same title/id.

You can also remove the test, it's a corner case.
Keep a commit next to the TryAdd explaining the compromise.

**Comment by comwanga** [comment]
Understood. I’ll simplify the change to use TryAdd, add a comment explaining that duplicate receipt keys are intentionally ignored to prevent checkout from failing, and remove the corner-case test and its related imports

File: BTCPayServer/Services/Invoices/PosReceiptData.cs [hunk 1]
```diff
@@ -87,7 +87,8 @@ void UpdateFromCart(IEnumerable<AppItem> appItems, PosAppData jposData, string c
             var totalPrice = displayFormatter.Currency(cartItem.Price * cartItem.Count, currency, DisplayFormatter.CurrencyFormat.Symbol);
             var ident = selectedChoice.Title ?? selectedChoice.Id;
             var key = selectedChoice.PriceType == AppItemPriceType.Fixed ? ident : $"{ident} ({singlePrice})";
-            cartData.Add(key, $"{cartItem.Count} x {singlePrice} = {totalPrice}");
+            // Duplicate receipt keys are intentionally ignored to prevent checkout from failing.
+            cartData.TryAdd(key, $"{cartItem.Count} x {singlePrice} = {totalPrice}");
```
[the remark this answers is not in the document]
Comment (comwanga) [reply BTCPayServer/Services/Invoices/PosReceiptData.cs:91]: This behavior is intentional following the maintainer discussion above. The accepted compromise is that later receipt entries with the same generated key may be omitted, while the complete cart still determines the invoice total. Using TryAdd prevents the key collision from failing checkout and keeps the implementation intentionally simple.

**Comment by comwanga** [comment]
Thanks for the clear guidance. I’ve made the requested changes. The implementation now uses `TryAdd` with a comment explaining the intentional compromise, and I removed the disambiguation logic, corner-case test and related imports. I also verified the simplified fix locally, and checkout completes successfully without the duplicate-key exception.
